src/cairoUtilFuncGen/abi/indexEncode.ts

```diff
@@ -0,0 +1,427 @@
+import assert from 'assert';
+import {
+  ArrayType,
+  generalizeType,
+  SourceUnit,
+  StructDefinition,
+  TypeNode,
+  UserDefinedType,
+} from 'solc-typed-ast';
+import { AST } from '../../ast/ast';
+import { printTypeNode } from '../../utils/astPrinter';
+import { CairoType, MemoryLocation, TypeConversionContext } from '../../utils/cairoTypeSystem';
+import { TranspileFailedError } from '../../utils/errors';
+import {
+  getElementType,
+  getPackedByteSize,
+  isAddressType,
+  isDynamicallySized,
+  isDynamicArray,
+  isStruct,
+  safeGetNodeType,
+} from '../../utils/nodeTypeProcessing';
+import { uint256 } from '../../warplib/utils';
+import { delegateBasedOnType, mul } from '../base';
+import { MemoryReadGen } from '../memory/memoryRead';
+import { AbiBase, removeSizeInfo } from './base';
+
+const IMPLICITS =
+  '{bitwise_ptr : BitwiseBuiltin*, range_check_ptr : felt, warp_memory : DictAccess*}';
+
+/**
+ * It is a special class used for encoding of indexed arguments in events.
+ * More info at:
+   https://docs.soliditylang.org/en/v0.8.14/abi-spec.html#encoding-of-indexed-event-parameters
+ */
+export class IndexEncode extends AbiBase {
+  protected override functionName = 'index_encode';
+  protected memoryRead: MemoryReadGen;
+
+  constructor(memoryRead: MemoryReadGen, ast: AST, sourceUnit: SourceUnit) {
+    super(ast, sourceUnit);
+    this.memoryRead = memoryRead;
+  }
+
+  public getOrCreate(types: TypeNode[]): string {
+    const key = types.map((t) => t.pp()).join(',');
+    const existing = this.generatedFunctions.get(key);
+    if (existing !== undefined) {
+      return existing.name;
+    }
+
+    const [params, encodings] = types.reduce(
+      ([params, encodings], type, index) => {
+        const cairoType = CairoType.fromSol(type, this.ast, TypeConversionContext.Ref);
+        params.push({ name: `param${index}`, type: cairoType.toString() });
+        encodings.push(
+          // padding is not required for strings and bytes
+          this.generateEncodingCode(type, 'bytes_index', `param${index}`, false),
+        );
+        return [params, encodings];
+      },
+      [new Array<{ name: string; type: string }>(), new Array<string>()],
+    );
+
+    const cairoParams = params.map((p) => `${p.name} : ${p.type}`).join(', ');
+    const funcName = `${this.functionName}${this.generatedFunctions.size}`;
+    const code = [
+      `func ${funcName}${IMPLICITS}(${cairoParams}) -> (result_ptr : felt){`,
+      `  alloc_locals;`,
+      `  let bytes_index : felt = 0;`,
+      `  let (bytes_array : felt*) = alloc();`,
+      ...encodings,
+      `  let (max_length256) = felt_to_uint256(bytes_index);`,
+      `  let (mem_ptr) = wm_new(max_length256, ${uint256(1)});`,
+      `  felt_array_to_warp_memory_array(0, bytes_array, 0, mem_ptr, bytes_index);`,
+      `  return (mem_ptr,);`,
+      `}`,
+    ].join('\n');
```

Comment:
I'm just following the standard that rodrigo has used for abiEncode impls.

Reply:
multline template string may introduce extra whitespaces at the beginning that we want to avoid. And even identation would be bad 
```js
        node.canonicalSignature(ABIEncoderVersion.V2),
      ),
      ...keysInsertions,
`   // keys: pack 31 byte felts into a single 248 bit felt,
   let (keys_len: felt, keys: felt*) = pack_bytes_felt(${BYTES_IN_FELT_PACKING}, ${BIG_ENDIAN}, keys_len, keys);,
   // data arrays,
   let data_len: felt = 0;,
   let (data: felt*) = alloc();`,
      ...dataInsertions,
      `   // data: pack 31 bytes felts into a single 248 bits felt`,
      `   let (data_len: felt, data: felt*) = pack_bytes_felt(${BYTES_IN_FELT_PACKING}, ${BIG_ENDIAN}, data_len, data);`,
      `   emit_event(keys_len, keys, data_len, data);`,
      `   return ();`,
      `}`,
```

Reply:
Good point. We could think of using https://www.npmjs.com/package/endent or https://www.npmjs.com/package/common-tags to fix the issue.

Reply:
we can do it in a separate PR. Because we've to change the template string standard throughout the code base. can you create a notion ticket for it?

Reply:
I agree, but I don't want to be hasty. First we have to discuss that with @rodrigo-pino if he has established the convention :) I don't want to arbitrarily decide on code style conventions

Reply:
I'll create notion ticket if Rodrigo also agrees that is the preferred solution

Reply:
@piwonskp  @RohitRanjangit  I did it like that because I was following the existing  conventions too :sweat_smile: , `cairoUtilFuncWriter`  is fully written like this.

I don't mind changing it if it would improve readability. My only two drawbacks are the need for an extra library and that there is a lot to change :hot_face: 

Also, I find it pretty readable as it is, but I've done considerable work in this area and I am surely biased.